### PR TITLE
Generalize the data structures for the cells adjacent to an edge.

### PR DESCRIPTION
**Pull request by bangerth**
In 2d, only two cells can be adjacent to an edge, and the current data structures in the
mesh reorientation algorithm reflect this: each edge had 2 slots for adjacent cells, of which
zero, one, or two could be used at any given time.

On the other hand, in 3d, an arbitrary number of cells may be adjacent to an edge. I eventually
want to extend the algorithm re-written in #3166 to the 3d case as well. This patch generalizes
the data structure representing the cells adjacent to an edge by making it generic. The patch
only implements the 2d case, but the 3d case will be easy to add later on, and it makes the 'Edge'
data structure generic.

This passes all tests with 'grid' in their name. That's all I can easily run on my laptop.

### CODE REVIEW

**Comment by bangerth** [comment]
I'm pushing a second patch that accelerates a function. It is independent from the first one, but conveniently reviewed jointly if you want.

File: source/grid/grid_reordering.cc [hunk 1]
```diff
@@ -257,36 +338,10 @@ namespace internal
       OrientationStatus orientation_status;
 
       /**
-       * A structure that stores how this edge relates to the at most
-       * two cells that are next to it.
+       * Store the set of cells adjacent to this edge (these cells then
+       * also store *where* in the cell the edge is located.
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: missing ")"

File: source/grid/grid_reordering.cc [hunk 2]
```diff
@@ -146,8 +146,111 @@ namespace internal
     };
 
 
-    template <int dim>
-    class Edge;
+    /**
+     * A structure that store the index of a cell and, crucially, how a
+     * given edge relates this cell.
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: "relates *to* this cell"?

**Comment by bangerth** [comment]
Comments addressed, new version uploaded.

**Comment by tjhei** [comment]
:+1:

**Comment by bangerth** [comment]
Part of #3235.

**Pull request merged**